src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/BaseAsyncStream.cs

```diff
@@ -0,0 +1,118 @@
+        {
+            if (buffer == null)
+            {
+                throw new ArgumentNullException(nameof(buffer));
+            }
+
+            if ((uint)offset > buffer.Length)
+            {
+                throw new ArgumentOutOfRangeException(nameof(offset));
+            }
+
+            if ((uint)count > buffer.Length - offset)
+            {
+                throw new ArgumentOutOfRangeException(nameof(count));
+            }
+        }
+
+        /// <summary>
+        /// Validate the arguments to CopyTo, as would Stream.CopyTo, but with knowledge that
+        /// the source stream is always readable and so only checking the destination.
+        /// </summary>
+        protected static void ValidateCopyToArgs(Stream source, Stream destination, int bufferSize)
+        {
+            if (destination == null)
+            {
+                throw new ArgumentNullException(nameof(destination));
+            }
+
+            if (bufferSize <= 0)
+            {
+                throw new ArgumentOutOfRangeException(nameof(bufferSize), bufferSize, SR.ArgumentOutOfRange_NeedPosNum);
+            }
+
+            if (!destination.CanWrite)
+            {
+                throw destination.CanRead ?
+                    new NotSupportedException(SR.NotSupported_UnwritableStream) :
+                    (Exception)new ObjectDisposedException(nameof(destination), SR.ObjectDisposed_StreamClosed);
+            }
+        }
+
+        public sealed override int Read(byte[] buffer, int offset, int count)
+        {
+            ValidateBufferArgs(buffer, offset, count);
+            return ReadAsync(new Memory<byte>(buffer, offset, count), CancellationToken.None).GetAwaiter().GetResult();
+        }
+
+        public sealed override Task<int> ReadAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
+        {
+            ValidateBufferArgs(buffer, offset, count);
+            return ReadAsync(new Memory<byte>(buffer, offset, count), cancellationToken).AsTask();
+        }
+
+        public sealed override void Write(byte[] buffer, int offset, int count)
+        {
+            ValidateBufferArgs(buffer, offset, count);
+            WriteAsync(new Memory<byte>(buffer, offset, count), CancellationToken.None).GetAwaiter().GetResult();
+        }
+
+        public sealed override Task WriteAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
+        {
+            ValidateBufferArgs(buffer, offset, count);
+            return WriteAsync(new ReadOnlyMemory<byte>(buffer, offset, count), cancellationToken).AsTask();
+        }
+
+        public sealed override void Flush() => FlushAsync().GetAwaiter().GetResult();
+
+        public sealed override void CopyTo(Stream destination, int bufferSize) =>
+            CopyToAsync(destination, bufferSize, CancellationToken.None).GetAwaiter().GetResult();
+
+
+        //
+        // Methods which must be implemented by derived classes
+        //
```

Comment:
And FlushAsync, no?

Reply:
The base implementation of FlushAsync is a no-op, so it doesn't strictly need to be overridden (and isn't for the read streams). I could change this (force all derivations to implement FlushAsync) if you think it's worth it.

Reply:
Which base implementation is a nop?  Stream.FlushAsync is not a nop: it queues a work item that calls Flush.

Reply:
Hmmm, my mistake, sorry. I'll add FlushAsync here.